### PR TITLE
[2/Z] Connect User to DB

### PR DESCRIPTION
**Pull request by dash102**
This PR:
1. Takes out any dependence on user IDs
2. Adds profile update servlet and related necessities to allow servlet to function
3. Adds frontend for profile updates

What it functionally did:
Writes to and reads from DB in order to show actual user profile, and allows updates from the user (which will be reflected in the profile page and DB)

Additionally, it depends on 1) a local plugin not in this PR and 2) an Event method that will be written in the analogous Event PR

**I will re-style the code once it gets nearer to merging, but it is functionally ready for review.**

### CODE REVIEW

File: project/src/main/java/com/google/sps/utilities/DatabaseWrapper.java [hunk 1]
```diff
@@ -20,51 +27,58 @@ public DatabaseWrapper(String instanceId, String databaseId) {
...
-  public void insertUser(User user) {
+  public void insertOrUpdateUser(User user) {
+    // Given a user, update its corresponding row's new fields in DB
     SpannerOptions options = SpannerOptions.newBuilder().build();
     Spanner spanner = options.getService();
     DatabaseId db = DatabaseId.of(options.getProjectId(), instanceId, databaseId);
     DatabaseClient dbClient = spanner.getDatabaseClient(db);
 
-    List<Mutation> mutations = getMutationsFromBuilder(newInsertBuilderFromUser(), user);
+    List<Mutation> mutations = getMutationsFromBuilder(Mutation.newInsertOrUpdateBuilder(USER_TABLE), user);
```
Comment (yicyao) [inline project/src/main/java/com/google/sps/utilities/DatabaseWrapper.java:41]: I have a similar function for Events -- can you rename 'getMutationsFromBuilder' to 'getUserMutationsFromBuilder'

File: project/src/main/java/com/google/sps/servlets/UserProfileUpdateServlet.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        .add("interests", createJsonArray(user.getInterests()))
+        .add("skills", createJsonArray(user.getSkills()))
+        .build()
+        .toString();
+    response.setContentType("application/json;");
+    response.setCharacterEncoding("UTF-8");
+    response.getWriter().println(userJson);
+  }
+
+  public JsonArray createJsonArray(Set<String> elements) {
```
Comment (liangqigong) [inline project/src/main/java/com/google/sps/servlets/UserProfileUpdateServlet.java:41]: Does this need to be public? Maybe use private static?

File: project/src/main/java/com/google/sps/utilities/DatabaseWrapper.java [hunk 3]
```diff
@@ -20,51 +27,58 @@ public DatabaseWrapper(String instanceId, String databaseId) {
...
-   * Given a user, insert a row with all available fields into the DB
+   * Given a user, insert or update a row with all available fields into the DB
    *
-   * @param user the user to be inserted; user's ID field should not exist in DB
+   * @param user the user to be updated; user's email may or may not exist in DB
    */
-  public void insertUser(User user) {
+  public void insertOrUpdateUser(User user) {
+    // Given a user, update its corresponding row's new fields in DB
     SpannerOptions options = SpannerOptions.newBuilder().build();
```
Comment (liangqigong) [inline project/src/main/java/com/google/sps/utilities/DatabaseWrapper.java:36]: I am thinking maybe you can inject dbClient, so you don't need to create it in every function. 

You don't need to change this in the PR. Just a thought.

File: project/src/main/java/com/google/sps/utilities/DatabaseWrapper.java [hunk 4]
```diff
@@ -20,51 +27,58 @@ public DatabaseWrapper(String instanceId, String databaseId) {
...
+  public User readUserFromEmail(String email) {
+    // Given an email, return the corresponding user from the DB
     SpannerOptions options = SpannerOptions.newBuilder().build();
     Spanner spanner = options.getService();
     DatabaseId db = DatabaseId.of(options.getProjectId(), instanceId, databaseId);
     DatabaseClient dbClient = spanner.getDatabaseClient(db);
+    ResultSet resultSet =
+        dbClient
+            .singleUse()
+            .executeQuery(Statement.of(String.format("SELECT Name, Interests, Skills, EventsHosting, EventsParticipating, EventsVolunteering FROM %s WHERE Email='%s'", USER_TABLE, email)));
```
Comment (liangqigong) [inline project/src/main/java/com/google/sps/utilities/DatabaseWrapper.java:55]: The line is too long. Try to reformat it?

File: project/src/main/java/com/google/sps/utilities/DatabaseWrapper.java [hunk 5]
```diff
@@ -20,51 +27,58 @@ public DatabaseWrapper(String instanceId, String databaseId) {
...
 
-    List<Mutation> mutations = getMutationsFromBuilder(newUpdateBuilderFromUser(), user);
-    dbClient.write(mutations);
-    spanner.close();
-  }
+    if (!resultSet.next()) {
+      return new User.Builder(/* name = */ "anonymous", /* email = */ email).build();
+    }
```
Comment (liangqigong) [inline project/src/main/java/com/google/sps/utilities/DatabaseWrapper.java:59]: Use Optional here.
Return Optional.empty()

File: project/src/main/java/com/google/sps/utilities/DatabaseWrapper.java (hunk 4 above)
Comment (dash102) [reply project/src/main/java/com/google/sps/utilities/DatabaseWrapper.java:55]: Will do. Formatting is a bit wonky throughout but I'll include this fix with my other formatting fixes.

File: project/src/main/java/com/google/sps/utilities/Constants.java [hunk 6]
```diff
@@ -0,0 +1,6 @@
+package com.google.sps.utilities;
```
Comment (nikhitashanker) [inline project/src/main/java/com/google/sps/utilities/Constants.java:1]: Maybe we can call this class  DatabaseConstants?

File: project/src/main/java/com/google/sps/utilities/DatabaseWrapper.java (hunk 1 above)
Comment (dash102) [reply project/src/main/java/com/google/sps/utilities/DatabaseWrapper.java:41]: Done

File: project/src/main/java/com/google/sps/servlets/UserProfileUpdateServlet.java (hunk 2 above)
Comment (dash102) [reply project/src/main/java/com/google/sps/servlets/UserProfileUpdateServlet.java:41]: Done, moved it to the CommonUtils class because I expect we will need it several times as a helper anyways.

File: project/src/main/java/com/google/sps/utilities/DatabaseWrapper.java (hunk 3 above)
Comment (dash102) [reply project/src/main/java/com/google/sps/utilities/DatabaseWrapper.java:36]: Ack, might be an application-wide change if we have time, as we have many of these types of methods from different PRs.

File: project/src/main/java/com/google/sps/utilities/Constants.java (hunk 6 above)
Comment (dash102) [reply project/src/main/java/com/google/sps/utilities/Constants.java:1]: Done

File: project/src/main/java/com/google/sps/utilities/Constants.java [hunk 7]
```diff
@@ -0,0 +1,6 @@
...
+
+public final class Constants {
+  public static final String DB_INSTANCEID = "step-188-instance";
+  public static final String DB_DATABASEID = "event-organizer-db";
```
Comment (nikhitashanker) [inline project/src/main/java/com/google/sps/utilities/Constants.java:5]: nit: change to DATABASE_ID INSTANCE_ID

File: project/src/main/java/com/google/sps/utilities/DatabaseWrapper.java (hunk 5 above)
Comment (dash102) [reply project/src/main/java/com/google/sps/utilities/DatabaseWrapper.java]: Done

File: project/src/main/java/com/google/sps/utilities/Constants.java (hunk 7 above)
Comment (dash102) [reply project/src/main/java/com/google/sps/utilities/Constants.java]: Done

File: .github/workflows/lint.yaml [hunk 8]
```diff
@@ -0,0 +1,23 @@
+name: Lint
```
Comment (liangqigong) [inline .github/workflows/lint.yaml:1]: Could you move this to another PR?

File: project/src/main/java/com/google/sps/servlets/UserProfileUpdateServlet.java [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+/** Servlet that handles user profile edits. */
+@WebServlet("/profile-update")
+public class UserProfileUpdateServlet extends HttpServlet {
+  private static final String email =
+      UserServiceFactory.getUserService().getCurrentUser().getEmail();
+  private DatabaseWrapper databaseWrapper =
+      new DatabaseWrapper(DatabaseConstants.INSTANCE_ID, DatabaseConstants.DATABASE_ID);
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
```
Comment (liangqigong) [inline project/src/main/java/com/google/sps/servlets/UserProfileUpdateServlet.java:28]: Add comment for the public functions?

File: project/src/main/java/com/google/sps/utilities/CommonUtils.java [hunk 10]
```diff
@@ -1,9 +1,21 @@
 package com.google.sps.utilities;
```
Comment (liangqigong) [inline project/src/main/java/com/google/sps/utilities/CommonUtils.java:1]: Add test for the public functions fo the util class.

File: .github/workflows/lint.yaml (hunk 8 above)
Comment (dash102) [reply .github/workflows/lint.yaml]: Done

File: project/src/main/java/com/google/sps/servlets/UserProfileUpdateServlet.java (hunk 9 above)
Comment (dash102) [reply project/src/main/java/com/google/sps/servlets/UserProfileUpdateServlet.java:29]: Done

File: project/src/main/java/com/google/sps/utilities/CommonUtils.java (hunk 10 above)
Comment (dash102) [reply project/src/main/java/com/google/sps/utilities/CommonUtils.java:1]: Done

**Comment by dash102** [comment]
Ready for review

File: project/src/main/java/com/google/sps/data/User.java [hunk 11]
```diff
@@ -183,8 +172,9 @@ public String getEmail() {
     return this.eventsHosting;
   }
 
-  public Set<Long> getEventsHostingIds() {
-    Set<Long> eventIds = new HashSet<>();
+  /** Returns a Set of eventHosting Id fields */
```
Comment (liangqigong) [inline project/src/main/java/com/google/sps/data/User.java:175]: Actually this comment is not necessary here since the function name explain the purpose of the function.

Comment is needed for servlet functions because the name doesn't explain the function.

File: project/src/main/java/com/google/sps/data/User.java [hunk 12]
```diff
@@ -183,8 +172,9 @@ public String getEmail() {
...
+  public Set<String> getEventsHostingIds() {
+    Set<String> eventIds = new HashSet<>();
     for (Event event : this.eventsHosting) {
       eventIds.add(event.getId());
     }
```
Comment (liangqigong) [inline project/src/main/java/com/google/sps/data/User.java:180]: Try this?

eventHosting.stream().map(Event::getId).collect(Collectors.toSet());

File: project/src/main/java/com/google/sps/servlets/UserProfileUpdateServlet.java [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+          Json.createObjectBuilder()
+              .add("name", "anonymous")
+              .add("email", email)
+              .add("interests", CommonUtils.createJsonArray(new HashSet<>()))
+              .add("skills", CommonUtils.createJsonArray(new HashSet<>()))
+              .build()
+              .toString();
+    }
+    User user = userOptional.get();
+    String userJson =
```
Comment (liangqigong) [inline project/src/main/java/com/google/sps/servlets/UserProfileUpdateServlet.java:42]: Else?

Comment (liangqigong) [reply project/src/main/java/com/google/sps/servlets/UserProfileUpdateServlet.java:42]: BTW, also leave a TODO to add test for this.

File: project/src/main/java/com/google/sps/data/User.java (hunk 11 above)
Comment (dash102) [reply project/src/main/java/com/google/sps/data/User.java]: Done, that makes sense.

File: project/src/main/java/com/google/sps/data/User.java (hunk 12 above)
Comment (dash102) [reply project/src/main/java/com/google/sps/data/User.java]: Neat, thanks!

File: project/src/main/java/com/google/sps/servlets/UserProfileUpdateServlet.java (hunk 13 above)
Comment (dash102) [reply project/src/main/java/com/google/sps/servlets/UserProfileUpdateServlet.java]: Ah, missed the else - done.

File: project/src/main/java/com/google/sps/servlets/UserProfileUpdateServlet.java [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** Servlet that handles user profile edits. */
+@WebServlet("/profile-update")
+public class UserProfileUpdateServlet extends HttpServlet {
+  private static final String email =
+      UserServiceFactory.getUserService().getCurrentUser().getEmail();
+  private DatabaseWrapper databaseWrapper =
```
Comment (hlz03) [inline project/src/main/java/com/google/sps/servlets/UserProfileUpdateServlet.java:24]: private static final?

Comment (dash102) [reply project/src/main/java/com/google/sps/servlets/UserProfileUpdateServlet.java]: Ah, missed that - done

File: project/src/main/webapp/profile-edit.js [hunk 15]
```diff
@@ -0,0 +1,13 @@
+$(document).ready(function() {
+  fetch('/profile-update')
+      .then((response) => response.json())
```
Comment (nikhitashanker) [inline project/src/main/webapp/profile-edit.js:3]: Can use async and await style here to avoid chaining.

Comment (nikhitashanker) [reply project/src/main/webapp/profile-edit.js:3]: extract into an an async function then you can do 
const response = await fetch('/profile-update');
const userData = await response.json();

Comment (dash102) [reply project/src/main/webapp/profile-edit.js]: Done, thanks
